### ISSUE TITLE
Installing as a module from PyPI leaves installations vulnerable to bugs in dependencies

### ISSUE THREAD

**Issue by richvdh** [matrix-org/synapse#13315]
One of our [recommended](https://matrix-org.github.io/synapse/latest/setup/installation.html#installing-as-a-python-module-from-pypi) installation methods suggests installing Synapse as a module from PyPI.

This works fine, but if one of our downstream dependencies has a bug (such as a security vulnerability) then it is difficult for administrators to know they need to update their installation. This in turn leads us to have to increase our minimum dependency versions (see https://github.com/matrix-org/synapse/pull/13172 for example), which makes things difficult for downstream packagers.

**Comment by richvdh** [comment]
> If we do this, I think people will end up pip installing as long as there is a matrix-synapse on PyPI.

Yes, but the first problem here is that we are actually *recommending* an approach that leaves people in a mess. It's entirely reasonable to address this in two steps (where the second step may or may not happen any time soon):
 *  first, stop recommending a broken approach
 * second, worry about people who follow other approaches that we don't recommend.

**Comment by richvdh** [comment]
> Ideas for solving this:
> 
>     * stop recommending `pip` as a way to install synapse, and support only docker/debs
> 
>     * have poetry export a `requirements.txt` and get users to use it on upgrade somehow
> 
>     * decide that we no longer want to support downstream packagers and just pin to a single version of all our dependencies

It seems surprising to me that there is no poetry-blessed way of doing this.

**Comment by richvdh** [comment]
> stop recommending `pip` as a way to install synapse, and support only docker/debs

We'd better come up with a decent answer to https://github.com/matrix-org/synapse/issues/9944 if this is the route we want to take.

**Comment by DMRobertson** [comment]
It might be instructive to look for other standaloneish applications distributed on PyPI and see what they do to handle this situation. I'm struggling to think of a good example though. (Jupyter, black, isort, mypy all sprang to mind, but they have to coexist with devtools so I imagine they have lax requirements.)

**Comment by richvdh** [comment]
I think other comparable standaloneish applications simply aren't distributed via PyPI, and people are expected to use docker images or os-specific packages. Which amounts to my first suggestion: "stop recommending pip as a way to install synapse, and support only docker/debs".

**Comment by sbiberhofer** [comment]
Just to chime in here from a distribution packager's perspective: At least on FreeBSD we basically repackage the PyPI release, so locking dependencies on the released package would probably force us to move to the source release. We can, of course, always switch over and build our packages from there, but a minor heads-up would be appreciated if you go that route as that release doesn't provide a legacy setup.py that we currently rely on for reasons completely unrelated to the issue at hand.

**Comment by DMRobertson** [comment]
> as that release doesn't provide a legacy setup.py

`poetry-core` generates setup.py for us when we build sdists and wheels. [`python -m build`](https://github.com/pypa/build) should be all you need to generate an identical setup.py from source.

**Comment by gsemet** [comment]
We distribute a locked requirements.txt but it is not easily scalable. Need to be able to generate a -frozen version of the wheel. Should be easily done by repackaging the wheel I guess.

**Comment by gsemet** [comment]
We switched now to packaging our application using zipapp. This only requires a local interpreter. This is good solution to avoid dependency nightmare, and ensuring our user uses the same version of the dependencies we have carefully validated.
